File: solution/java/682.棒球比赛.java
import java.util.*;
/*
 * @lc app=leetcode.cn id=682 lang=java
 *
 * [682] 棒球比赛
 */

// @lc code=start
class Solution {
    public int calPoints(String[] ops) {
        int ret = 0, result = 0, n = 0;
        List<Integer> arr = new ArrayList<>();
        for (String s : ops) {
            n = arr.size();
            switch (s.charAt(0)) {
                case '+':
                    result = arr.get(n - 1) + arr.get(n - 2);
                    ret += result;
                    arr.add(result);
                    break;
                case 'D':
                    result = arr.get(n - 1) * 2;
                    ret += result;
                    arr.add(result);
                    break;
                case 'C':
                    result = arr.get(n - 1);
                    ret -= result;
                    arr.remove(n - 1);
                    break;
                default:
                    result = Integer.parseInt(s);
                    ret += result;
                    arr.add(result);
                    break;
            }
        }
        return ret;
    }
}
// @lc code=end

File: solution/python/682.棒球比赛.py
#
# @lc app=leetcode.cn id=682 lang=python3
#
# [682] 棒球比赛
#

import sys
import os

sys.path.append(os.path.join(os.path.dirname(__file__), ".."))

from typing import *
from common.node import *


# @lc code=start
class Solution:
    def calPoints(self, operations: List[str]) -> int:
        scores = []
        for operation in operations:
            if operation == "+":
                scores.append(scores[-1] + scores[-2])
            elif operation == "D":
                scores.append(scores[-1] * 2)
            elif operation == "C":
                scores.pop()
            else:
                scores.append(int(operation))
        return sum(scores)


# @lc code=end


if __name__ == "__main__":
    solution = Solution()
    test_cases = [
        (solution.calPoints, (["5", "2", "C", "D", "+"],), 30),
        (solution.calPoints, (["1", "C"],), 0),
    ]

    all_passed = True
    for idx, (func, args, expected) in enumerate(test_cases):
        try:
            result = func(*args)
            assert result == expected
            print(f"测试用例 {idx + 1} 通过: n = {args}, result = {result}")
        except AssertionError:
            all_passed = False
            print(
                f"测试用例 {idx + 1} 失败: n = {args}, 期望 = {expected}, 实际 = {result}"
            )

    file_path = os.path.basename(__file__).split(".")
    file_number = file_path[0]
    file_name = file_path[1]
    if all_passed:
        print(f'第 {file_number} 题 "{file_name}" 所有测试用例通过')
        sys.exit(0)
    else:
        print(f'第 {file_number} 题 "{file_name}" 部分测试用例失败')
        sys.exit(1)
